File: 0801-is-graph-bipartite/0801-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int[] color = new int[n];
        Arrays.fill(color, -1);
        for(int i = 0; i < n; i++){
            Queue<Integer> q = new LinkedList<>();
            if(color[i] != -1) continue;
            q.offer(i);
            while(!q.isEmpty()){
                int curr = q.poll();
                for(int it : graph[curr]){
                    if(color[it] == -1){
                        color[it] = 1 - color[curr];
                        q.offer(it);
                    }
                    else if(color[it] == color[curr]) return false;
                }
            }
        }
        return true;
    }
}

File: 0801-is-graph-bipartite/0801-is-graph-bipartite.cpp
class Solution {
private:
    bool dfs(vector<vector<int>>& graph, vector<int>& vis, int node, int color){
        vis[node] = color;
        for(auto it : graph[node]){
            if(vis[it] == -1){
                if(dfs(graph, vis, it, !color) == false) return false;
            }
            else if(vis[it] == color) return false;
        }
        return true;
    }
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> vis(n, -1);
        for(int i = 0; i < n; i++){
            if(vis[i] == -1){
                if(dfs(graph, vis, i, 0) == false) return false; 
            }
        }
        return true;
    }
};